### ISSUE TITLE
Rapid Outlier Detection

### ISSUE THREAD

**Issue by orgesleka** [scikit-learn-contrib/scikit-learn-contrib#25]
Request for project inclusion in scikit-learn-contrib

* Project name: rapid-outlier-detection
* Project description: Rapid Outlier Detectin via Sampling
* Authors: Orges Leka
* Current repository: https://github.com/orgesleka/rapid-outlier-detection
* Requirements:
 * [x] scikit-learn compatible (check_estimator passed)
 * [x] Documentation (guide, API reference, example gallery)
 * [x] Python3 compatible
 * [x ] PEP8 compliant

**Comment by agramfort** [comment]
where is the doc and gallery?

**Comment by orgesleka** [comment]
gallery: example_radid_outlier_detection.py
doc: is implemented as a docstring.

**Comment by agramfort** [comment]
it needs to be a web site we can review.

**Comment by ogrisel** [comment]
It would be great to reproduce some of the results of the benchmarks of the paper (kdd99, covertype, synthetic gaussian datasets) and compare with the methods already present in sklearn: IsolationForest, LOF and OneClassSVM.

**Comment by ogrisel** [comment]
Also please configure the continuous integration as done in the project template:

https://github.com/scikit-learn-contrib/project-template

**Comment by orgesleka** [comment]
Ok, thanks. I will do that, but give me some time as I am doing this on my free time.

**Comment by ogrisel** [comment]
No problem, there is no hurry. I read the paper and it's indeed very surprising that such a simple baseline seems to dominate all of the more complex and slower state of the art methods (on average).

**Comment by orgesleka** [comment]
Dear Olivier ( @ogrisel ). I wrote a small script to benchmark the algorithms IsolationForest, LOF and OneClassSVM on the datasets covtype, kdd1999, gaussian1e3. Prof. Mahito Sugiyama was so kind to send me the dataset. He also send the other synthetic gaussian dataset, but it is too large (over 1.4 GB)
If you could have a look on the script, and share your opinion on what might be improved, that would be nice.
Before you run the script benchmark.py do not forget to unzip the datasets.zip in the data/. folder.

I only ran the script for covtype (lof, rapid outlier) , gaussian1e3 (all four algorithms), as I get a memory error otherwise.

**Comment by ogrisel** [comment]
For the synthetic datasets you can re-synthesize them yourself.

**Comment by orgesleka** [comment]
still don't have enough memory on my pc to hold 1.4 GB and run the algorithms.

**Comment by ogrisel** [comment]
Can you please report your results the run you did? Could you reproduce some of the scores of the paper?

Please create an example folder with an updated version of https://github.com/ngoix/scikit-learn/blob/master/examples/covariance/plot_outlier_detection.py that includes your new method.

Also add another example script that compares fast methods (e.g. yours and IsolationForest) on medium scale datasets such as `sklearn.datasets.fetch_covtype` and `sklearn.datasets.fetch_kddcup99`. In particular it would be interesting to make a plot that shows the influence of the hyperparameters (e.g. `n_samples`) on the training time, testing time (e.g. to compute the outlier scores on 10000 test samples) and the test accuracy of the model. Same for isolation forest.

For the benchmark script, please do not do host specific hacks such as:

```python
sys.path.insert(1,'/usr/local/lib/python2.7/dist-packages/')
```

Instead, write a minimalist `setup.py` for your project and install it with `pip install --editable .`
 so that you can import the development version of your project from anywhere.

Also do not put megabytes size data in the source repo. Instead, either use the data fetchers from scikit-learn (e.g. `sklearn.datasets.fetch_covtype`, `sklearn.datasets.kddcup99`), generate synthetic Gaussian data with `np.random.RandomState(42).normal` or `multivariate_normal`. If you some benchmark dataset are not available in `sklearn.datasets`, write a helper script to download it in the local folder using `urllib.request.urlretrieve` under Python 3 (and optionally with `urllib.urlretrieve` under Python 2).

Also please configure travis CI for your repo to automatically launch the tests both for Python 2.7 and Python 3.6 (unless you don't won't to support Python 2 anymore and in which case only testing on Python 3 is fine but then you should make it explicit in the README.md). Also add the travis badge.

**Comment by agramfort** [comment]
you have 2 benchmarks in master:

https://github.com/scikit-learn/scikit-learn/blob/master/benchmarks/bench_lof.py
https://github.com/scikit-learn/scikit-learn/blob/master/benchmarks/bench_isolation_forest.py

you can reuse

**Comment by orgesleka** [comment]
@agramfort thank you.

**Comment by ogrisel** [comment]
Just for the record, this other recent method looks very promising as well: [Linear-time Outlier Detection via Sensitivity](https://arxiv.org/abs/1605.00519). Linear-time and more accurate on many benchmarks than (although the individual benchmark AUPRC values reported for the Rapid Outlier Detection method in both papers seem to differ quite significantly).
